cheroot/connections.py

```diff
@@ -49,6 +50,71 @@ def prevent_socket_inheritance(sock):
         fcntl.fcntl(fd, fcntl.F_SETFD, old_flags | fcntl.FD_CLOEXEC)
 
 
+class _ThreadsafeSelector:
+    """Threadsafe wrapper around a DefaultSelector.
+
+    There are 2 thread contexts in which it may be accessed:
+    * the selector thread
+    * one of the worker threads in workers/threadpool.py
+
+    The expected read/write patterns are:
+    * iter: selector thread
+    * register: selector thread and threadpool, via put()
+    * unregister: selector thread only
+
+    Notably, this means _ThreadsafeSelector never needs to worry
+    that connections will be removed behind its back.
+
+    The lock is held when iterating or modifying the selector,
+    but is not required when select()ing on it.
+    """
+
+    def __init__(self):
+        self._selector = selectors.DefaultSelector()
+        self._lock = threading.Lock()
+
+    def __len__(self):
+        # NOTE: the _SelectorMapping returned by get_map()
+        # implements its own __len__ routine which does not
+        # iterate the map, so it is not required
+        # to hold the lock here.
+        return len(self._selector.get_map())
+
+    @property
+    def connections(self):
+        """Retrieve connections registered with the selector."""
+        with self._lock:
+            mapping = self._selector.get_map() or {}
+            for _, (_, sock_fd, _, conn) in mapping.items():
+                yield (sock_fd, conn)
+
+    def register(self, fileobj, events, data=None):
+        """Register fileobj with the selector."""
+        with self._lock:
+            return self._selector.register(fileobj, events, data)
+
+    def unregister(self, fileobj):
+        """Unregister fileobj from the selector."""
```

Comment:
And "unregister" should probably be whitelisted @ https://github.com/cherrypy/cheroot/blob/master/docs/spelling_wordlist.txt.